### PR TITLE
gcd harmony

### PR DESCRIPTION
**Pull request by SansPapyrus683**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

### CODE REVIEW

File: solutions/gold/kattis-gcdharmony.mdx [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+### Refinement
+
+As stated in the hints, just keeping values from 2 to 100 won't work.
+The maximum cost of making a tree harmonious is $2 \cdot 5000=10^4$,
+so we have to keep values all the way up til that.
+
+How do we transition, then, with such large states?
+
+Notice first that it would never make sense for a node to have a final value with a prime factor greater than 100.
```
Comment (ryanchou-dev) [inline solutions/gold/kattis-gcdharmony.mdx:57]: elaborate?

File: solutions/gold/kattis-gcdharmony.mdx [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+			res = std::min(res, min_cost[0][i]);
+		}
+		return res;
+	}
+};
+
+int main() {
+	int node_num;
+	std::cin >> node_num;
+	assert(2 * node_num <= MAX_POSS);
```
Comment (ryanchou-dev) [inline solutions/gold/kattis-gcdharmony.mdx:177]: you can probably remove input validation (?)

**Review by ryanchou-dev — COMMENTED** [review]
code looks amazing as always 😍

File: solutions/gold/kattis-gcdharmony.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply solutions/gold/kattis-gcdharmony.mdx:177]: fair enough

File: solutions/gold/kattis-gcdharmony.mdx [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+int main() {
+	int node_num;
+	std::cin >> node_num;
+	vector<int> vals(node_num);
+	for (int &v : vals) { std::cin >> v; }
+	vector<vector<int>> neighbors(node_num);
+	for (int i = 0; i < node_num - 1; i++) {
+		int a, b;
+		std::cin >> a >> b;
+		neighbors[--a].push_back(--b);
```
Comment (TheGamingMousse) [inline solutions/gold/kattis-gcdharmony.mdx:185]: I'd personally write `--a, --b;` on a separate line (but it doesn't really matter lol)

File: solutions/gold/kattis-gcdharmony.mdx [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+		neighbors[b].push_back(a);
+	}
+
+	cout << GCDTree(neighbors, vals).min_adjustment() << endl;
```
Comment (TheGamingMousse) [inline solutions/gold/kattis-gcdharmony.mdx:189]: it obv doesn't matter in this case, but i think (generally speaking) `"\n"` is preferable to `endl`

File: solutions/gold/kattis-gcdharmony.mdx [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+#include <iostream>
+#include <map>
+#include <vector>
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+constexpr int MAX_VAL = 100;   // the largest value a node can have
+const int MAX_POSS = 2 * 5e3;  // defined as 2 * (max # of nodes)
```
Comment (TheGamingMousse) [inline solutions/gold/kattis-gcdharmony.mdx:96]: why make the first one constexpr but the second one const?

Comment (SansPapyrus683) [reply solutions/gold/kattis-gcdharmony.mdx:96]: good catch

File: solutions/gold/kattis-gcdharmony.mdx (hunk 4 above)
Comment (SansPapyrus683) [reply solutions/gold/kattis-gcdharmony.mdx:189]: uh actually i think endl is better bc it flushes the buffer?
i only use \n in a loop but
point is most of the code in the guide here uses endl now

File: solutions/gold/kattis-gcdharmony.mdx (hunk 3 above)
Comment (SansPapyrus683) [reply solutions/gold/kattis-gcdharmony.mdx:185]: uh i've used this in multiple other places before, so there is precedent

File: solutions/gold/kattis-gcdharmony.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/gold/kattis-gcdharmony.mdx:96]: decided to make both const due to precedent

File: solutions/gold/kattis-gcdharmony.mdx (hunk 3 above)
Comment (TheGamingMousse) [reply solutions/gold/kattis-gcdharmony.mdx:185]: oh ok lol

File: solutions/gold/kattis-gcdharmony.mdx (hunk 4 above)
Comment (TheGamingMousse) [reply solutions/gold/kattis-gcdharmony.mdx:189]: hmmm ok

**Review by brebenelmihnea — APPROVED** [review]

**Pull request merged**